broker-core/src/main/java/io/zeebe/broker/workflow/processor/instance/CreateWorkflowInstanceProcessor.java

```diff
@@ -0,0 +1,210 @@
+    if (!setVariablesFromDocument(controller, record, workflowInstanceKey)) {
+      return;
+    }
+
+    final ElementInstance workflowInstance =
+        createElementInstance(workflow, workflowInstanceKey, record.getVariables());
+    streamWriter.appendFollowUpEvent(
+        workflowInstanceKey,
+        WorkflowInstanceIntent.ELEMENT_ACTIVATING,
+        workflowInstance.getValue());
+    controller.accept(
+        WorkflowInstanceCreationIntent.CREATED, record.setInstanceKey(workflowInstanceKey));
+  }
+
+  private boolean isValidWorkflow(
+      CommandControl<WorkflowInstanceCreationRecord> controller, DeployedWorkflow workflow) {
+    if (workflow.getWorkflow().getNoneStartEvent() == null) {
+      controller.reject(RejectionType.INVALID_STATE, ERROR_MESSAGE_NO_NONE_START_EVENT);
+      return false;
+    }
+
+    return true;
+  }
+
+  private boolean setVariablesFromDocument(
+      CommandControl<WorkflowInstanceCreationRecord> controller,
+      WorkflowInstanceCreationRecord record,
+      long workflowInstanceKey) {
+    try {
+      variablesState.setVariablesLocalFromDocument(workflowInstanceKey, record.getVariables());
+    } catch (MsgpackReaderException e) {
+      Loggers.WORKFLOW_PROCESSOR_LOGGER.error(ERROR_INVALID_VARIABLES_LOGGED_MESSAGE, e);
+      controller.reject(
+          RejectionType.INVALID_ARGUMENT,
+          String.format(ERROR_INVALID_VARIABLES_REJECTION_MESSAGE, e.getMessage()));
+
+      return false;
+    }
+
+    return true;
+  }
+
+  private ElementInstance createElementInstance(
+      DeployedWorkflow workflow, long workflowInstanceKey, DirectBuffer variables) {
+    newWorkflowInstance.reset();
+    newWorkflowInstance.setBpmnProcessId(workflow.getBpmnProcessId());
+    newWorkflowInstance.setVersion(workflow.getVersion());
+    newWorkflowInstance.setWorkflowKey(workflow.getKey());
+    newWorkflowInstance.setWorkflowInstanceKey(workflowInstanceKey);
+    newWorkflowInstance.setBpmnElementType(BpmnElementType.PROCESS);
+    newWorkflowInstance.setElementId(workflow.getWorkflow().getId());
+    newWorkflowInstance.setFlowScopeKey(-1);
+    newWorkflowInstance.setPayload(variables); // todo: to be removed in follow up issues (#1858)
+
+    final ElementInstance instance =
+        elementInstanceState.newInstance(
+            workflowInstanceKey, newWorkflowInstance, WorkflowInstanceIntent.ELEMENT_ACTIVATING);
+    elementInstanceState.flushDirtyState();
+
+    return instance;
+  }
+
+  private DeployedWorkflow getWorkflow(
+      WorkflowInstanceCreationRecord record, CommandControl controller) {
+    final DeployedWorkflow workflow;
+
+    if (record.getKey() >= 0) {
+      workflow = getWorkflow(record.getKey(), controller);
+    } else {
+      final DirectBuffer bpmnProcessId = record.getBpmnProcessId();
+
+      if (bpmnProcessId.capacity() == 0) {
+        controller.reject(RejectionType.INVALID_ARGUMENT, ERROR_MESSAGE_NO_IDENTIFIER_SPECIFIED);
+        workflow = null;
```

Comment:
Why null workflow here and not on declaration?

Reply:
So I can have a final variable (i.e. safer, compiler enforced single assignment). Having a single place to assign a return value, and a single return statement, just usually means its easier to think about possible cases (e.g. cyclomatic complexity)